src/abstractdataframe/abstractdataframe.jl

```diff
@@ -1535,29 +1546,94 @@ julia> vcat(df1, df3, cols=:intersect)
+   5 │      2        5        5  missing
+   6 │      2        6        6  missing
+   7 │      3        7  missing        7
+   8 │      3        8  missing        8
+   9 │      3        9  missing        9
+  10 │      4  missing  missing  missing
+
+julia> vcat(df1, df2, df3, df4, cols=:union, source=:source => 'a':'d')
+10×4 DataFrame
+ Row │ source  A        B        C
+     │ Char    Int64?   Int64?   Int64?
+─────┼───────────────────────────────────
+   1 │ a             1        1  missing
+   2 │ a             2        2  missing
+   3 │ a             3        3  missing
+   4 │ b             4        4  missing
+   5 │ b             5        5  missing
+   6 │ b             6        6  missing
+   7 │ c             7  missing        7
+   8 │ c             8  missing        8
+   9 │ c             9  missing        9
+  10 │ d       missing  missing  missing
 ```
 """
 Base.vcat(dfs::AbstractDataFrame...;
           cols::Union{Symbol, AbstractVector{Symbol},
-                      AbstractVector{<:AbstractString}}=:setequal) =
-    reduce(vcat, dfs; cols=cols)
+                      AbstractVector{<:AbstractString}}=:setequal,
+          source::Union{Nothing, SymbolOrString,
+                        Pair{<:SymbolOrString, <:AbstractVector}}=nothing) =
+    reduce(vcat, dfs; cols=cols, source=source)
 
 function Base.reduce(::typeof(vcat),
                      dfs::Union{AbstractVector{<:AbstractDataFrame},
                                 Tuple{AbstractDataFrame, Vararg{AbstractDataFrame}}};
                      cols::Union{Symbol, AbstractVector{Symbol},
-                     AbstractVector{<:AbstractString}}=:setequal)
-    return _vcat(AbstractDataFrame[df for df in dfs if ncol(df) != 0]; cols=cols)
+                                 AbstractVector{<:AbstractString}}=:setequal,
+                     source::Union{Nothing, SymbolOrString,
+                                   Pair{<:SymbolOrString, <:AbstractVector}}=nothing)
+    if source === nothing
+        return _vcat(AbstractDataFrame[df for df in dfs if ncol(df) != 0]; cols=cols)
+    else
+        len = length(dfs)
+        if source isa SymbolOrString
+            col, vals = source, 1:len
+        else
+            @assert source isa Pair{<:SymbolOrString, <:AbstractVector}
+            col, vals = source
+        end
+
+        idx = findfirst(df -> columnindex(df, col) > 0, dfs)
+        if idx !== nothing
+            throw(ArgumentError("source column name exists in data frame " *
+                                " passed in position $idx"))
+        end
+
+        if len != length(vals)
+            throw(ArgumentError("number of passed source column identifiers " *
+                                "does not match the number of passed columns"))
+        end
+
+        dfs′ = Vector{AbstractDataFrame}(undef, len)
+        for (i, (v, df)) in enumerate(zip(vals, dfs))
+            dfs′[i] = insertcols!(copy(df, copycols=false), 1, col => Ref(v))
+        end
```

Comment:
Probably not a big deal, but wouldn't it be more efficient (and not really more complex) to create the column only after concatenating?

Reply:
This is a design decision related with an empty data frame. As you can see in tests and examples, if you pass `DataFrame()` to `vcat` now a single row is created with all `missing` values except indicator column. Otherwise we would drop such a data frame.

But maybe dropping it is preferable. What do you think?

Reply:
I have thought about it and concluded that it is better to drop it. I will change the implementation.